src/math/mat4.h

```diff
@@ -0,0 +1,536 @@
+        ConstexprSwap(m_MatrixElements[6],  m_MatrixElements[9]);
+        ConstexprSwap(m_MatrixElements[7],  m_MatrixElements[13]);
+        ConstexprSwap(m_MatrixElements[11], m_MatrixElements[14]);
+    }
+
+    constexpr Mat4 Mat4::Transpose(const Mat4& matrix) noexcept {
+        Mat4 copy = matrix;
+        copy.Transpose();
+        return copy;
+    }
+
+    constexpr Mat4 Mat4::Identity() noexcept {
+        return Mat4(1.0f);
+    }
+
+    constexpr Mat4 Mat4::Translation(const Vec3& translation) noexcept {
+        Mat4 result(1.0f);
+
+        result.m_MatrixElements[0 + 3 * 4] = translation.GetX();
+        result.m_MatrixElements[1 + 3 * 4] = translation.GetY();
+        result.m_MatrixElements[2 + 3 * 4] = translation.GetZ();
+
+        return result;
+    }
+
+    constexpr Mat4 Mat4::Scale(const Vec3& scale) noexcept {
+        Mat4 result(1.0f);
+
+        result.m_MatrixElements[0 + 0 * 4] = scale.GetX();
+        result.m_MatrixElements[1 + 1 * 4] = scale.GetY();
+        result.m_MatrixElements[2 + 2 * 4] = scale.GetZ();
+
+        return result;
+    }
+
+    constexpr Mat4 Mat4::Rotation(float angle, const Vec3& axis) noexcept {
+        Mat4 result(1.0f);
+
+        float angleInRadians = angle / 180.0f * static_cast<float>(M_PI);
+        float x = axis.GetX();
+        float y = axis.GetY();
+        float z = axis.GetZ();
+        float c = static_cast<float>(cos(angleInRadians));
+        float s = static_cast<float>(sin(angleInRadians));
+        float omc = 1 - c;
+
+        result.m_MatrixElements[0 + 0 * 4] = x * x * omc + c;
+        result.m_MatrixElements[1 + 0 * 4] = x * y * omc + z * s;
+        result.m_MatrixElements[2 + 0 * 4] = x * z * omc - y * s;
+
+        result.m_MatrixElements[0 + 1 * 4] = x * y * omc - z * s;
+        result.m_MatrixElements[1 + 1 * 4] = y * y * omc + c;
+        result.m_MatrixElements[2 + 1 * 4] = y * z * omc + x * s;
+
+        result.m_MatrixElements[0 + 2 * 4] = x * z * omc + y * s;
+        result.m_MatrixElements[1 + 2 * 4] = y * z * omc - x * s;
+        result.m_MatrixElements[2 + 2 * 4] = z * z * omc + c;
+
+        return result;
+    }
+
+    constexpr Mat4
+    Mat4::Orthographic(float left, float right, float bottom, float top, float near, float far) noexcept {
+        Mat4 result(1.0f);
+
+        result.m_MatrixElements[0 + 0 * 4] = 2.0f / (right - left);
+        result.m_MatrixElements[1 + 1 * 4] = 2.0f / (top - bottom);
+        result.m_MatrixElements[2 + 2 * 4] = 2.0f / (near - far);
+
+        result.m_MatrixElements[0 + 3 * 4] = (right + left) / (left - right);
+        result.m_MatrixElements[1 + 3 * 4] = (top + bottom) / (bottom - top);
+        result.m_MatrixElements[2 + 3 * 4] = (far + near) / (near - far);
+
+        return result;
+    }
+
+    constexpr Mat4 Mat4::Perspective(float fov, float aspectRatio, float near, float far) noexcept {
+        Mat4 result(0.0f);
+
+        float fovInRadians = fov / 180.0f * static_cast<float>(M_PI);
```

Comment:
Next place were same thing is implemented

Reply:
Corrected :+1: